File: MinFallingPath.py
## Problem2 (https://leetcode.com/problems/minimum-falling-path-sum/)

# Time Complexity : O(n^2)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

class Solution(object):
    def minFallingPathSum(self, A):
        if not A:
            return 0
        n = len(A)
        for i in range(1, n):
            for j in range(n):
                if j == 0:
                    A[i][j] += min(A[i - 1][j], A[i - 1][j + 1])
                elif j == n - 1:
                    A[i][j] += min(A[i - 1][j], A[i - 1][j - 1])
                else:
                    A[i][j] += min(A[i - 1][j], A[i - 1][j + 1], A[i - 1][j - 1])
        return min(A[-1])

File: min_falling_path.cpp
// Time Complexity : O(n*n)
// Space Complexity :  O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this - 

class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        
        int len = matrix.size();
        int i, j;
        int min_val;
        
        for(i = 1; i < len; i++)
        {
            for(j = 0; j < len; j++)
            {
                min_val = INT_MAX;
                
                if(j-1 >= 0)
                {
                    if(matrix[i-1][j-1] < min_val)
                    {
                        min_val = matrix[i-1][j-1];
                    }
                }
                
                if(matrix[i-1][j] < min_val)
                {
                    min_val = matrix[i-1][j];
                }
                
                if(j+1 < len)
                {
                    if(matrix[i-1][j+1] < min_val)
                    {
                        min_val = matrix[i-1][j+1];
                    }
                }
                cout << min_val << endl;
                matrix[i][j] += min_val;
            }
        }
        
        min_val = INT_MAX;
        
        for(i = 0; i < len; i++)
        {
            cout << matrix[len-1][i] << endl;
            if(matrix[len-1][i] < min_val)
            {
                min_val = matrix[len-1][i];
            }
        }
        
        return min_val;
    }
};